Add Anycubic Kobra 3 Max

robertbaker opened the pull request:
# Description
Adds Kobra 3 Max now that the latest firmware officially supports OrcaSlicer.
Ports the printer and profile files from AnycubicSlicerNext to OrcaSlicer.


Closes #9507 
Depends on #12455 

# Screenshots/Recordings/Graphs
N/A

## Tests

Did a build with a clean config folder.
added printer

robertbaker committed:
Add Anycubic Kobra 3 Max
Fix Kobra 3 Max printable area and JSON spacing

rene-bayer commented:
Not able to get a print started with connected ace
Without any adjustments, the printer asks for color assignment (will upload screenshots later)

When adding:
```
; ams_info = begin
; paint_info = [{"material_type":"PETG","paint_color":[30,30,30],"paint_index":0}]
; project_info = {"flush_multiplier":1.0,"flush_volumes_matrix":[0.0,131.0,131.0,131.0,131.0,0.0,131.0,131.0,131.0,131.0,0.0,131.0,131.0,131.0,131.0,0.0],"flush_volumes_vector":
; ams_info = end
```

to the end of the gcode file, it at least lets me start the print, but then gets stuck at "Preperation:Parsing"
<img width="290" height="497" alt="image" src="https://github.com/user-attachments/assets/eb6096f3-7ddd-4ac0-8dbe-d3a9bbbfcc8e" />

real-limitless commented:
@rene-bayer and @robertbaker is there a way I can test this as well to help you folks contribute to this pull request?

rene-bayer commented:
@robertbaker 
The generated gcode will not print on current firmware, as long as these lines find themselves in the generated gcode files:
```
; filament_adaptive_volumetric_speed = 1,0,0,0
; filament_adhesiveness_category = 0,0,0,0
; filament_change_length = 10,10,10,10
; filament_colour_type = 1;1;1;1
```

I have absolutely no idea, why this lands in the gcodes.

rene-bayer commented:
These lines are not added when I slice the files with 2.3.1 and the same profile which I created with 2.3.2-dev

SoftFever commented:
> Have a minor issue related to the bed exclusion area. I was able to slice and send to the printer but it threw an error
> 
> <img alt="image" width="1150" height="748" src="https://private-user-images.githubusercontent.com/51850403/523021227-dee1982b-78d8-473b-aa84-fcd42ea422a6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYwMTgzMDcsIm5iZiI6MTc2NjAxODAwNywicGF0aCI6Ii81MTg1MDQwMy81MjMwMjEyMjctZGVlMTk4MmItNzhkOC00NzNiLWFhODQtZmNkNDJlYTQyMmE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE4VDAwMzMyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmNGI1OGVjNDYwZDQ0ZDEyMDhlNDM2N2RkNGM1ZjAyOTc0MTVmNTE4NzRhZmM1MjY5NzA1ODFmMDY1ZjVmYTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AvBPJWvMDnsgLlKAr8-091SQZ9hcN1P2PJvIdNeR7ck">

Can you try adding a working `bed_exclude_area`gcode in the "Machine start G-code" and see if that works?  
If this turns out to be the issue, I can look into adding multi-polygon support for `bed_exclude_area`.

I also tried to check their OrcaSlicer fork to understand how they handle `bed_exclude_area`, but it looks like they violated the GPL and didn’t open-source their fork.

rene-bayer commented:
> We received a response from the software department that, due to the busy development schedule, perfect compatibility with Orcaslicer cannot be guaranteed at present. 
We will consider making some targeted optimizations in the future.
Thank you for supporting Anycubic.
have a nice day!
Kind regards,
ANYCUBIC Team

thelegendtubaguy commented:
I think this should wait until #12455 is merged as it likely solves the problem here as well.


This PR on top of main (resolving conflicts) with @Mock1 's post processing script works.  The exported gcode works and the printer will happily let you select filament and start prints.
<img width="6144" height="8160" alt="PXL_20260417_211951897" src="https://github.com/user-attachments/assets/86a1e4f4-5cdc-41fd-8c59-696238982773" />
<img width="6144" height="8160" alt="PXL_20260417_211946470" src="https://github.com/user-attachments/assets/ef56fece-0064-4cb6-a1b8-367fb2624763" />

Edit: Something is still funky with the printable area/bed exclude area.  If I place a small object on the bed that definitely should be allowed, it will let me slice and print no problem.  However, I cannot use auto arrange.  Auto arrange will kick the part off the bed as if it were too big for the bed.  Not sure why that's happening.

SoftFever commented:
Hi everyone,  
Sorry for the late response.  
I'll merge this PR first.  
For now, please use a post-processing script as @Mock1 suggested as a workaround.  
We will make this step easier soon. Until then, this is probably the only viable approach.  
It's not maintainable in OrcaSlicer to modify the core code to "fix" an issue caused by a printer manufacturer doing something funny with a standard G-code comment.  
Also, please try to encourage them to fix this obvious bug on their end.

> Just as an FYI, if anyone is wanting to get this working while a fix for the machine specific GCode exclusion is figured out; I was able to get my K3M printing with this, both in single color and multi color modes. I added a post-processing script to remove the GCode lines that were causing the firmware to hang. Now it's pretty much "normal" functionality. To be clear, this is not saying this is the answer. If the printer is available in the software, it should work without users needing to add any post-processing scripts. Or the script should be part of it. But this is a way to use the PR right now.
> 
> ```
> import sys, os, tempfile
> 
> bad_prefixes = [
>     "; filament_adaptive_volumetric_speed",
>     "; filament_adhesiveness_category",
>     "; filament_change_length",
>     "; filament_colour_type",
>     "; bed_temperature_formula = by_first_filament",
> ]
> 
> path = sys.argv[1]
> with open(path, 'r') as f:
>     lines = f.readlines()
> 
> tmp = path + ".tmp"
> with open(tmp, 'w') as f:
>     for line in lines:
>         if not any(line.startswith(p) for p in bad_prefixes):
>             f.write(line)
> 
> os.replace(tmp, path)
> ```

robertbaker commented:
An improved version of the script **required** at this time for the Kobra 3 Max.
Gist: https://gist.github.com/robertbaker/e4fce61e6adca4dce304e68265729a91

Copy either the python (mac, linux) or powershell (windows) script to a file in a path Orca can access.

Add this to OrcaSlicer > Process: Others tab > (scroll down to) Post Processing Scripts
`python3 my/path/to/k3m-fix.py`

```
import sys
import os

# Lines to remove (prefix match)
BAD_PREFIXES = [
    "; filament_adaptive_volumetric_speed",
    "; filament_adhesiveness_category",
    "; filament_change_length",
    "; filament_colour_type",
    "; bed_temperature_formula = by_first_filament",
]

# Pre-strip prefixes for faster matching
BAD_PREFIXES = tuple(p.lstrip() for p in BAD_PREFIXES)

def should_remove(line: str) -> bool:
    stripped = line.lstrip()
    return stripped.startswith(BAD_PREFIXES)

def main():
    if len(sys.argv) < 1:
        return 0

    path = sys.argv[1]
    tmp_path = path + ".tmp"

    changed = False

    with open(path, "r", encoding="utf-8", errors="ignore") as fin, \
         open(tmp_path, "w", encoding="utf-8") as fout:

        for line in fin:
            if should_remove(line):
                changed = True
                continue
            fout.write(line)

    # If nothing changed, delete tmp and exit cleanly
    if not changed:
        os.remove(tmp_path)
        return 0

    # Windows-safe overwrite
    try:
        os.remove(path)
    except FileNotFoundError:
        pass

    os.rename(tmp_path, path)
    return 0

if __name__ == "__main__":
    main()
```

For Windows users who don't want to deal with python:
Add this to OrcaSlicer > Process: Others tab > (scroll down to) Post Processing Scripts
`powershell -ExecutionPolicy Bypass -File "C:\path\to\k3m-fix.ps1"`

```
param(
    [string]$Path
)

# Prefixes to remove
$BadPrefixes = @(
    "; filament_adaptive_volumetric_speed",
    "; filament_adhesiveness_category",
    "; filament_change_length",
    "; filament_colour_type",
    "; bed_temperature_formula = by_first_filament"
)

# Temp file
$Tmp = "$Path.tmp"

# Track if anything changed
$Changed = $false

# Read + write safely
$in  = [System.IO.File]::OpenText($Path)
$out = New-Object System.IO.StreamWriter($Tmp, $false, [System.Text.Encoding]::UTF8)

try {
    while (($line = $in.ReadLine()) -ne $null) {
        $trimmed = $line.TrimStart()

        $match = $false
        foreach ($prefix in $BadPrefixes) {
            if ($trimmed.StartsWith($prefix)) {
                $match = $true
                break
            }
        }

        if (-not $match) {
            $out.WriteLine($line)
        } else {
            $Changed = $true
        }
    }
}
finally {
    $in.Close()
    $out.Close()
}

# If nothing changed, delete tmp and exit
if (-not $Changed) {
    Remove-Item $Tmp -Force
    exit 0
}

# Windows-safe overwrite
Remove-Item $Path -Force
Rename-Item $Tmp $Path -Force

exit 0
```
